experimental/AlgebraicShifting/src/PartialShiftGraph.jl

```diff
@@ -0,0 +1,297 @@
+"""
+isless_lex(K1::ComplexOrHypergraph, K2::ComplexOrHypergraph) = isless_lex(Set(facets(K1)), Set(facets(K2)))
+
+@doc raw"""
+    partial_shift_graph_vertices(F::Field,::SimplicialComplex, W::Union{WeylGroup, Vector{WeylGroupElem}};)
+
+Given a field `F` discover the vertices of the partial shift graph starting from `K`
+using exterior partial shifts corresponding to elements in `W`.
+Returns a `Vector{SimplicialCompplex}` ordered lexicographically.
+
+#Example
+```jldoctest
+julia> K = simplicial_complex([[1, 2], [2, 3], [3, 4]])
+Abstract simplicial complex of dimension 1 on 4 vertices
+
+julia> shifts = partial_shift_graph_vertices(QQ, K, weyl_group(:A, 3))
+6-element Vector{SimplicialComplex}:
+ Abstract simplicial complex of dimension 1 on 4 vertices
+ Abstract simplicial complex of dimension 1 on 4 vertices
+ Abstract simplicial complex of dimension 1 on 4 vertices
+ Abstract simplicial complex of dimension 1 on 4 vertices
+ Abstract simplicial complex of dimension 1 on 4 vertices
+ Abstract simplicial complex of dimension 1 on 4 vertices
+
+julia> facets.(shifts)
+6-element Vector{Vector{Set{Int64}}}:
+ [Set([3, 1]), Set([4, 1]), Set([2, 1])]
+ [Set([3, 1]), Set([2, 3]), Set([2, 1]), Set([4])]
+ [Set([3, 1]), Set([4, 2]), Set([2, 1])]
+ [Set([4, 3]), Set([3, 1]), Set([2, 1])]
+ [Set([2, 1]), Set([2, 3]), Set([4, 2])]
+ [Set([2, 1]), Set([2, 3]), Set([4, 3])]
+```
+"""
+function partial_shift_graph_vertices(F::Field,
+                                      K::SimplicialComplex,
+                                      W::Union{WeylGroup, Vector{WeylGroupElem}};)
+  current = K
+  visited = [current]
+  phi = isomorphism(PermGroup, parent(first(W)))
+  # by properties of algebraic shifting
+  # we know that K will be the last in this sorted list
+  # sorting here should also speed up unique according to julia docs
+  unvisited = unique(
+    x -> Set(facets(x)),
+    sort([exterior_shift(F, K, phi(w)) for w in W]; lt=isless_lex))[1:end - 1]
+
+  while(!isempty(unvisited))
+    current = pop!(unvisited)
+    push!(visited, current)
+    shifts = unique(
+      x -> Set(facets(x)),
+      sort([exterior_shift(F, current, phi(w)) for w in W]; lt=isless_lex))[1:end - 1]
+
+    # dont visit things twice
+    new_facets = filter(x -> !(x in Set.(facets.(visited))), Set.(facets.(shifts)))
+    unvisited = simplicial_complex.(
+      union(Set.(facets.(unvisited)), new_facets))
+  end
+  return sort(collect(visited); lt=isless_lex)
+end
+
+""" Compute the multi edges, that is, for each complex `K`  compute which
+    other complexes can be reached by applying the partial shifts `deltas`
+    to K, and store the shift matrices that give rise to each edge."""
+function multi_edges(F::Field,
+                     permutations::Vector{PermGroupElem},
+                     complexes::Vector{Tuple{Int,T}},
+                     complex_labels::Dict{Set{Set{Int}}, Int}
+) :: Dict{Tuple{Int, Int}, Vector{PermGroupElem}}  where T <: ComplexOrHypergraph;
```

Comment:
Also the `:: Dict{Tuple{Int, Int}, Vector{PermGroupElem}}` is a conversion annotation, i.e. Julia will insert a call to `convert(Dict{Tuple{Int, Int}, Vector{PermGroupElem}}, RESULT)` where `RESULT` is whatever this function returns. This is different from a type annotations. Just double checking that this is indeed what you intended.

Reply:
I see there's a problem. I am sure that this function returns this type, but of course I don't want to silently convert when I change the function in the future. Going without any annotation does not convince me, either, because my impression is that this partly clarifies what this (somewhat obscure) function does. Plus, my linter profits from it, but I guess that's no argument.

Reply:
In this case, I think you could just move this to the return statement. Because there it will be a type annotation aka julia will throw an Error if the types don't match (instead of silently converting here)

Reply:
I've put type annotations on the return statements in these cases